### PR TITLE
[prompter] Add echoless passphrase prompt

### PR DESCRIPTION
**Pull request by townsend2010**
This allows entering the passphrase for `set local.passphrase` and `register` at the terminal without echo.

### CODE REVIEW

File: src/client/common/prompters.cpp [hunk 1]
```diff
@@ -20,20 +20,76 @@
...
     term->cout() << text << ": ";
 
-    std::string value;
-    std::getline(term->cin(), value);
-
-    if (!term->cin().good())
-        throw PromptException("Failed to read value");
+    auto value = get_input(term->cin());
 
     return value;
```
Comment (Saviq) [inline src/client/common/prompters.cpp:45]:
```suggestion
    return value = get_input(term->cin());
```

File: include/multipass/cli/prompters.h [hunk 2]
```diff
@@ -55,6 +55,23 @@ class PlainPrompter : public BasePrompter
...
+    std::string prompt(const std::string&) const override;
+};
+
+class NewPassphrasePrompter : public PassphrasePrompter
+{
+public:
+    using PassphrasePrompter::PassphrasePrompter;
+    using PassphrasePrompter::prompt;
+
+    std::string prompt(const std::string&, const std::string&) const;
```
Comment (Saviq) [inline include/multipass/cli/prompters.h:73]: Not sold on having a special interface here :/

Maybe we need to repurpose the argument to `prompt()` to be a piece of the actual prompt printed, and then the prompter has a standard format for it?

File: src/client/cli/cmd/register.cpp [hunk 3]
```diff
@@ -62,26 +64,52 @@ QString cmd::Register::description() const
...
 
-    request.set_passphrase(parser->positionalArguments().first().toStdString());
+    if (parser->positionalArguments().empty())
+    {
+        if (!term->is_live())
+        {
+            cerr << "The terminal is not live: The passphrase argument is required\n";
+            return ParseCode::CommandLineError;
+        }
+
```
Comment (Saviq) [inline src/client/cli/cmd/register.cpp:89]:
```suggestion
```

User could have piped it in, for whatever reason?

File: src/client/common/prompters.cpp [hunk 4]
```diff
@@ -20,20 +20,76 @@
...
+        term->cout() << "\n" << prompt2;
+
+        // Confirm the passphrase is the same by re-entering it
+        if (passphrase == get_input(term->cin()))
+        {
+            break;
+        }
+        else
+        {
+            term->cout() << "\nPassphrases do not match. Please try again.\n";
```
Comment (Saviq) [inline src/client/common/prompters.cpp:88]: I think we should just exit…

**Review by Saviq — CHANGES_REQUESTED** [review]
Some tweaks inline.

File: src/client/common/prompters.cpp (hunk 1 above)
Comment (townsend2010) [reply src/client/common/prompters.cpp]: Done

File: include/multipass/cli/prompters.h (hunk 2 above)
Comment (townsend2010) [reply include/multipass/cli/prompters.h]: Ok, I refactored this quite a bit.  Let me know what you think.

File: src/client/cli/cmd/register.cpp (hunk 3 above)
Comment (townsend2010) [reply src/client/cli/cmd/register.cpp]: Yep, done.

File: src/client/common/prompters.cpp (hunk 4 above)
Comment (townsend2010) [reply src/client/common/prompters.cpp]: Ok, done.

**Review by Saviq — CHANGES_REQUESTED** [review]
Should we `set_cin_echo(true)` in `~UnixTerminal()` to be safe?

I've not looked through the tests properly yet, but wanted to drop these in.

File: src/client/cli/cmd/register.cpp [hunk 5]
```diff
@@ -64,26 +66,46 @@ QString cmd::Register::description() const
 
 mp::ParseCode cmd::Register::parse_args(mp::ArgParser* parser)
 {
-    parser->addPositionalArgument("passphrase", "Passphrase to register with the Multipass service", "<passphrase>");
+    parser->addPositionalArgument("passphrase",
+                                  "Trusted passphrase to send to the Multipass service. If omitted, a prompt will be "
+                                  "displayed for entering the passphrase.",
+                                  "<passphrase>");
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/client/cli/cmd/register.cpp:72]: Oh, right, it's now optional with this change :slightly_smiling_face:

**Comment by townsend2010** [comment]
> Should we `set_cin_echo(true)` in `~UnixTerminal()` to be safe?

I guess it wouldn't hurt to do that.  At worst, it would basically be a no-op.

File: src/client/cli/cmd/register.cpp [hunk 6]
```diff
@@ -64,26 +66,46 @@ QString cmd::Register::description() const
 
 mp::ParseCode cmd::Register::parse_args(mp::ArgParser* parser)
 {
-    parser->addPositionalArgument("passphrase", "Passphrase to register with the Multipass service", "<passphrase>");
+    parser->addPositionalArgument("passphrase",
+                                  "Trusted passphrase to send to the Multipass service. If omitted, a prompt will be "
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/client/cli/cmd/register.cpp:70]: Ok, I'll keep the old wording and just add the `If omitted` part.  But yeah, we need to decide on the name of the command soon.

File: src/client/cli/cmd/register.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/client/cli/cmd/register.cpp]: Done.

**Comment by townsend2010** [comment]
The `Terminal` implementations now force echo to be on regardless of the current setting.

**Review by Saviq — CHANGES_REQUESTED** [review]

File: src/client/common/prompters.cpp [hunk 7]
```diff
@@ -20,20 +20,57 @@
...
 
-    std::string value;
-    std::getline(term->cin(), value);
+    return get_input(term->cin());
+}
+
+mp::PassphrasePrompter::PassphrasePrompter(Terminal* term) : PlainPrompter(term)
+{
+    term->set_cin_echo(false);
+}
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/client/common/prompters.cpp:49]: Yeah, very good point.

[the remark this answers is not in the document]
Comment (townsend2010) [reply src/client/common/prompters.cpp]: I've fixed this in hopefully a better manner.

**Comment by Saviq** [comment]
This needs a rebase, please.

**Comment by townsend2010** [comment]
> This needs a rebase, please.

Done.

File: tests/unix/test_unix_terminal.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
```
Comment (Saviq) [inline tests/unix/test_unix_terminal.cpp:2]:
```suggestion
 * Copyright (C) 2022 Canonical, Ltd.
```

File: tests/mock_terminal.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
```
Comment (Saviq) [inline tests/mock_terminal.h:2]:
```suggestion
 * Copyright (C) 2022 Canonical, Ltd.
```

File: tests/test_cli_client.cpp [hunk 10]
```diff
@@ -2393,10 +2399,65 @@ TEST_F(Client, registerCmdHelpOk)
...
+    std::istringstream cin;
+    mpt::MockTerminal mock_terminal;
+
+    cin.str(passphrase + "\n");
+
+    EXPECT_CALL(mock_terminal, cout).WillRepeatedly(ReturnRef(cout));
+    EXPECT_CALL(mock_terminal, cerr).WillRepeatedly(ReturnRef(cerr));
+    EXPECT_CALL(mock_terminal, cin).WillRepeatedly(ReturnRef(cin));
+    EXPECT_CALL(mock_terminal, set_cin_echo(false)).Times(1);
+    EXPECT_CALL(mock_terminal, set_cin_echo(true)).Times(1);
```
Comment (Saviq) [inline tests/test_cli_client.cpp:2415]: All these could go in a fixture, and true/false should go in a sequence?

File: tests/test_cli_prompters.cpp [hunk 11]
```diff
@@ -51,16 +52,91 @@ TEST_F(CLIPrompters, PlainReturnsText)
...
+    const std::string expected_output{"Please enter passphrase: \n"};
+    const std::string passphrase{"foo"};
+    mpt::MockTerminal mock_terminal;
+
+    cin.str(passphrase + "\n");
+
+    EXPECT_CALL(mock_terminal, set_cin_echo(false)).Times(1);
+    EXPECT_CALL(mock_terminal, set_cin_echo(true)).Times(1);
+    EXPECT_CALL(mock_terminal, cout()).WillRepeatedly([this]() -> std::ostream& { return cout; });
+    EXPECT_CALL(mock_terminal, cin()).WillOnce([this]() -> std::istream& { return cin; });
```
Comment (Saviq) [inline tests/test_cli_prompters.cpp:68]: Why not this?

```suggestion
    EXPECT_CALL(mock_terminal, cout()).WillRepeatedly(ReturnRef(cout));
    EXPECT_CALL(mock_terminal, cin()).WillOnce(ReturnRef(cin));
```

Below, too.

File: tests/test_cli_prompters.cpp [hunk 12]
```diff
@@ -51,16 +52,91 @@ TEST_F(CLIPrompters, PlainReturnsText)
...
+TEST_F(CLIPrompters, passphraseCallsEchoAndReturnsExpectedPassphrase)
+{
+    const std::string expected_output{"Please enter passphrase: \n"};
+    const std::string passphrase{"foo"};
+    mpt::MockTerminal mock_terminal;
+
+    cin.str(passphrase + "\n");
+
+    EXPECT_CALL(mock_terminal, set_cin_echo(false)).Times(1);
+    EXPECT_CALL(mock_terminal, set_cin_echo(true)).Times(1);
```
Comment (Saviq) [inline tests/test_cli_prompters.cpp:66]: Could use a sequence here.

**Review by Saviq — CHANGES_REQUESTED** [review]
Final tweaks to the tests.

File: tests/test_cli_prompters.cpp (hunk 11 above)
Comment (townsend2010) [reply tests/test_cli_prompters.cpp:68]: Right, not sure why I missed that....

File: tests/test_cli_prompters.cpp (hunk 12 above)
Comment (townsend2010) [reply tests/test_cli_prompters.cpp]: Done

File: tests/test_cli_prompters.cpp (hunk 11 above)
Comment (townsend2010) [reply tests/test_cli_prompters.cpp]: Done

File: tests/test_cli_client.cpp (hunk 10 above)
Comment (townsend2010) [reply tests/test_cli_client.cpp]: Yep, all done.

File: tests/mock_terminal.h (hunk 9 above)
Comment (townsend2010) [reply tests/mock_terminal.h]: Pushed

File: tests/unix/test_unix_terminal.cpp (hunk 8 above)
Comment (townsend2010) [reply tests/unix/test_unix_terminal.cpp]: Pushed

**Comment by townsend2010** [comment]
@Saviq,

Ping, how's this looking?

**Review by Saviq — APPROVED** [review]
Right!

All good :)

bors merge

**Pull request merged**